### DISCUSSION TITLE
Intermediate process before asking admin password

### DISCUSSION THREAD

**Discussion by buzunboy** [sparkle-project/Sparkle#2526] [Q&A]
## Summary

- Hi, we are using `.pkg` distribution for our Sparkle updates, and we would like to make things more automated. We use our custom drivers and directly using `SPUUpdater`, and the idea is we download automatically the updates and only inform the user once before we need to ask for admin password, which is happening when extraction starts. 

- Is it possible to add a new delegate method for `SPUUpdaterDelegate` which only continues to extraction if reply block is called, like below: 

## Possible Fix

```swift
- (void)downloadDriverDidDownloadUpdate:(SPUDownloadedUpdate *)downloadedUpdate
{
   ... 
    _resumableUpdate = downloadedUpdate;

    if (updater != nil && [updaterDelegate respondsToSelector:@selector(updater:shouldStartExtracting:invocation:)]) {
        [updaterDelegate updater:updater shouldStartExtracting:_updateItem invocation:^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self extractUpdate:downloadedUpdate];
            });
        }];
    } else {
        [self extractUpdate:downloadedUpdate];
    }
}
```
and with new method in the `SPUUpdaterDelegate.h`:
```swift
- (void)updater:(SPUUpdater *)updater shouldStartExtracting:(SUAppcastItem *)item invocation:(void (^)(void))reply;
```

So flow would be like:

1. User/App calls `SPUUpdater.checkForUpdates`
2. Sparkle calls `func showUpdateFound(with appcastItem: SUAppcastItem, state: SPUUserUpdateState, reply: @escaping (SPUUserUpdateChoice) -> Void)`
3. App automatically replies with `reply(.install)`
4. Sparkle starts & finishes the download, before extracting checks the implementation of `updater:shouldStartExtracting:invocation:` and only continues to extract if it's responded and invocation is called by the app.

I can create a pull request as well, posted as an issue to discuss first.

Thanks and best 

## Version

2.5.2, 2.6.0

**Comment by zorgiepoo** [comment]
Please see https://github.com/sparkle-project/Sparkle/discussions/2452
